Implement pod DELETE

tummychow opened the pull request:
please review code for sanity, not ready for merge. I haven't tested it or implemented intent safety checks.

tummychow commented:
okay this has safety checking for the preparer's pod ID in the intent set. ready for review

anthonybishopric commented on pkg/preparer/orchestrate.go:
```diff
@@ -84,22 +89,34 @@ func (p *Preparer) WatchForPodManifestsForNode(quitAndAck chan struct{}) {
+				// if the preparer's own ID is missing from the intent set, we
+				// assume it was damaged and discard it
+				if !checkResultsForID(intentResults, POD_ID) {
+					p.Logger.NoFields().Errorln("Intent results set did not contain p2-preparer pod ID")
```
Lets add a bit more color to this message - "... pod ID - it is possible that Consul's data is corrupt". This might help someone looking at preparer logs who isn't aware of this class of error.

tummychow replied:
will fix

anthonybishopric commented on pkg/preparer/orchestrate.go:
```diff
@@ -130,26 +147,37 @@ func (p *Preparer) handlePods(podChan <-chan pods.Manifest, quit <-chan struct{}
+				pod := pods.NewPod(nextLaunch.ID, pods.PodPath(p.podRoot, nextLaunch.ID))
 
-				ok := p.installAndLaunchPod(&manifestToLaunch, pod, manifestLogger)
+				ok := p.installAndLaunchPod(nextLaunch, pod, manifestLogger)
```
Lets move removal to a separate function from install, for clarity purposes

tummychow replied:
fair enough, but in that case I think the sha comparison function should also be factored out

anthonybishopric commented on pkg/preparer/result_set_test.go:
```diff
@@ -0,0 +1,34 @@
+		{Manifest: pods.Manifest{Id: "bar"}},
+	}
+
+	ValidatePairList(t, ZipResultSets(intent, reality))
```
I'd like this test to be a bit more precise: have the expected outcomes explicitly stated, ie "there's a pair with intent foo, but no reality foo", etc

tummychow replied:
yep I can add a bit more verification here

tummychow commented:
- rebased
- split up sha comparison, install/launch and halt/uninstall into separate functions

will embellish tests momentarily

Let's also talk about hooks on deletion

tummychow commented:
- tests for result set improved
- actually deletes the pod from reality now as well

tummychow commented:
added hooks for before_uninstall. i'll give it a quick spin to convince myself it works

petertseng commented:
well it lgtm, anyone else?

anthonybishopric commented on pkg/preparer/orchestrate.go:
```diff
@@ -176,92 +205,113 @@ func (p *Preparer) authorize(manifest pods.Manifest, logger logging.Logger) bool
-			p.tryRunHooks(hooks.AFTER_LAUNCH, pod, newManifest, logger)
-		}
-		return err == nil && ok
+	p.tryRunHooks(hooks.BEFORE_UNINSTALL, pod, pair.Reality, logger)
```
:+1:

anthonybishopric commented:
:shipit:
